src/lib/jsonParseV2/parseRegistration.cpp

```diff
@@ -0,0 +1,332 @@
+  }
+
+  if (dataProvided.ObjectEmpty())
+  {
+    *errorStringP = "/dataProvided/ is empty";
+    return false;
+  }
+
+  if (dataProvided.HasMember("entities"))
+  {
+    bool b = parseEntitiesVector(ciP, &dataProvidedP->entities, dataProvided["entities"], errorStringP);
+    if (b == false)
+    {
+      return false;
+    }
+  }
+
+  if (dataProvided.HasMember("attrs"))
+  {
+    if (!parseStringVector(&dataProvidedP->attributes, dataProvided["attrs"], "attrs", true, errorStringP))
+    {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+
+
+/* ****************************************************************************
+*
+* httpParse -
+*/
+static bool httpParse(ngsiv2::Http* httpP, const rapidjson::Value& jsonHttp, const std::string& fieldName, std::string* errorStringP)
+{
+  if (!jsonHttp.IsObject())
+  {
+    *errorStringP = "/" + fieldName + "/ must be a JSON object";
+    return false;
+  }
+
+  if (!jsonHttp.HasMember("url"))
+  {
+    *errorStringP = "/" + fieldName + "/ must contain a field /url/";
+    return false;
+  }
+
+  const rapidjson::Value& url = jsonHttp["url"];
+
+  if (!url.IsString())
+  {
+    *errorStringP = "/url/ field of /" + fieldName + "/ must be a JSON string";
+    return false;
+  }
+
+  httpP->url = url.GetString();
+
+  if (httpP->url == "")
+  {
+    *errorStringP = "/url/ field of /" + fieldName + "/ cannot be the empty string";
+    return false;
+  }
+
+  std::string  host;
+  int          port;
+  std::string  path;
+  std::string  protocol;
+
+  if (parseUrl(httpP->url, host, port, path, protocol) == false)
+  {
+    *errorStringP = "/url/ field of /" + fieldName + "/ is invalid";
+    return false;
+  }
+
+  return true;
+}
+
+
+
+/* ****************************************************************************
+*
+* providerParse -
+*/
+static bool providerParse(ConnectionInfo* ciP, ngsiv2::Provider* providerP, const rapidjson::Value& provider, std::string* errorStringP)
+{
+  bool b;
+
+  if (!provider.IsObject())
+  {
+    *errorStringP = "/provider/ must be a JSON object";
+    return false;
+  }
+
+  if (provider.ObjectEmpty())
+  {
+    *errorStringP = "/provider/ empty";
+    return false;
+  }
+
+  if (provider.HasMember("http"))
+  {
+    b = httpParse(&providerP->http, provider["http"], "http", errorStringP);
+    if (b == false)
+    {
+      return false;
+    }
+  }
+  else
+  {
+    *errorStringP = "/provider/ must contain a field /http/";
+    return false;
+  }
+
+  if (provider.HasMember("supportedForwardingMode"))
+  {
+    const rapidjson::Value& supportedForwardingModeValue = provider["supportedForwardingMode"];
+    std::string             supportedForwardingMode;
+
+    if (!supportedForwardingModeValue.IsString())
+    {
+      *errorStringP = "/supportedForwardingMode/ must be a JSON string";
+      return false;
+    }
+
+    supportedForwardingMode = supportedForwardingModeValue.GetString();
+
+    //
+    // Check Valid strings for supportedForwardingMode
```

Comment:
Only all is allowed at the present moment. The usage of any of the other modes should end with a Not Implemented error by the moment, generated at parsing stage (i.e. at this point in the code).

Reply:
In addition, the Not Implemented error generation should be covered by some .test

Reply:
See test case 35 in `v2_reg_create_errors.test`

Reply:
Looking to case 35, it is not the same. There you are using "invalid" as forwarding and checking you get BadRequest. What I mean is to use "query", "update", etc. and check you get "NotImplemented".

Reply:
ok, I understand.
But ... let's discuss this. Skype ...

Reply:
Fixed in 1a49367